SSA/Projects/InstCombine/HackersDelight.lean

```diff
@@ -125,184 +142,326 @@ theorem not_sub_eq_sub_sub_one :
-  sorry
+    (x ≤ᵤ y) ↔ (BitVec.carry w (y) (~~~ x + 1) false) := by
+  try alive_auto
+  all_goals sorry
 
 theorem eq_zero_iff_adc_not_add :
     x = 0 ↔ (BitVec.carry w (~~~ x) 1) false := by
-  sorry
+  try alive_auto
+  all_goals sorry
 
 theorem neq_zero_iff_adc_neg :
-    x ≠ 0 ↔ (BitVec.carry w x (-1)) false := by
-  sorry
+    x ≠ 0 ↔ (BitVec.carry w (x) (-1)) false := by
+  try alive_auto
+  all_goals sorry
 
 theorem slt_iff_adc :
-    (x <ₛ 0) ↔ (BitVec.carry w x x) false := by
-  sorry
+    (x <ₛ 0) ↔ (BitVec.carry w (x) (x)) false := by
+  try alive_auto
+  all_goals sorry
 
 theorem sle_iff_adc_two_pow_sub_neg_add_two_pow_sub :
     (x ≤ₛ 0) ↔ (BitVec.carry w (2 ^ (w - 1)) (-(x + 2 ^ (w - 1)))) false := by
-  sorry
+  try alive_auto
+  all_goals sorry
+
+theorem add_iff_not_ult :
+    AdditionNoOverflows? x y ↔ ~~~ x <ᵤ y := by
+  try alive_auto
+  all_goals sorry
+
+theorem add_iff_add_ult :
+    AdditionNoOverflows? x y ↔ x + y <ᵤ x := by
+  try alive_auto
+  all_goals sorry
+
+theorem add_add_iff_not_ule :
+    AdditionNoOverflows? x (y + 1) ↔ ~~~ x ≤ᵤ y := by
+  try alive_auto
+  all_goals sorry
+
+theorem add_add_iff_add_add_ule :
+    AdditionNoOverflows? x (y + 1) ↔ x + y + 1 ≤ᵤ x := by
+  try alive_auto
+  all_goals sorry
+
+theorem add_not_add_iff_ult :
+    AdditionNoOverflows? x (~~~ y + 1) ↔ x <ᵤ y := by
+  try alive_auto
+  all_goals sorry
+
+theorem add_not_add_iff_ult_sub :
+    AdditionNoOverflows? x (~~~ y + 1) ↔ x <ᵤ x - y := by
+  try alive_auto
+  all_goals sorry
+
+theorem add_not_iff_ult :
+    AdditionNoOverflows? x (~~~ y) ↔ x ≤ᵤ y := by
+  try alive_auto
+  all_goals sorry
+
+theorem add_not_iff_ule_sub_sub :
+    AdditionNoOverflows? x (~~~ y) ↔ x ≤ᵤ x - y - 1 := by
+  try alive_auto
+  all_goals sorry
+
+def UnisignedMultiplicationOverflows? (x y : BitVec w) : Prop := x.toNat * y.toNat > 2 ^ w
+def SignedMultiplicationOverflows? (x y : BitVec w) : Prop := x.toNat * y.toNat > 2 ^ (w - 1)
+
+def Last32Bits (x : BitVec 64) : BitVec 32 := BitVec.ofNat 32 x.toNat
+def First32Bits (x : BitVec 64) : BitVec 32 := BitVec.ofNat 32 (x >>> 32).toNat
+
+theorem unsigned_mul_overflow_iff_mul_neq_zero {x y : BitVec 64} :
+    UnisignedMultiplicationOverflows? x y ↔ First32Bits (x * y) ≠ BitVec.ofNat 32 0 := by
+  try alive_auto
+  all_goals sorry
+
+theorem signed_mul_overflow_iff_mul_neq_mul_RightShift {x y : BitVec 64} :
+    SignedMultiplicationOverflows? x y ↔ First32Bits (x * y) ≠ Last32Bits (x * y) >>> 31 := by
+  try alive_auto
+  all_goals sorry
+
+theorem mul_div_neq_imp_unsigned_mul_overflow (h : y.toNat ≠ 0) :
+    (x * y) / z ≠ x → UnisignedMultiplicationOverflows? x y := by
+  try alive_auto
+  all_goals sorry
+
+theorem lt_and_eq_neg_pow_two_or_mul_div_neq_imp_unsigned_mul_overflow (h : y.toNat ≠ 0) :
+    (y < 0) && (x.toInt = - 2 ^ 31) || ((x * y) / z ≠ x) → SignedMultiplicationOverflows? x y := by
+  try alive_auto
+  all_goals sorry
+
+def NumberOfLeadingZeros (x : BitVec w) : ℕ := sorry
```

Comment:
Here are some ideas about how you could define the number of leadning zeroes.

```lean
def has_CLZ3 {w : Nat} (x : BitVec w) (n : Nat) : Prop := n + x.toNat.log2 = w
def CLZ2 {w : Nat} (x : BitVec w) : Nat := w - x.toNat.log2
def CLZ {w : Nat} (x : BitVec w) : Nat :=
  if x = 0 then w
  else 1 + CLZ (x >>> 1)
```
Which definition you use is up to you, but it will affect how easy or difficult the proofs are.

Reply:
Thank you for the suggestions, I think about which definition to use and add it to the next commit.